### PR TITLE
Start documenting compatibility

### PR DESCRIPTION
**Pull request by czoido**
Closes: https://github.com/conan-io/docs/issues/2481

### CODE REVIEW

File: reference/conanfile/methods.rst [hunk 1]
```diff
@@ -1245,6 +1245,32 @@ in any other OS will be the standard one, as if the ``build_id()`` method was no
...
+should return a list of compatible configurations. For example, if we want that binaries
+built with gcc versions 4.8, 4.7 and 4.6 are considered compatible with the ones compiled
+with 4.9 we could declare the ``compatibility()`` like this:
+
+..  code-block:: python
+
+    def compatibility(self):
+        if self.settings.compiler == "gcc" and self.settings.compiler.version == "4.9":
+            return [{"settings": [("compiler.version", v)]}
+                    for v in ("4.8", "4.7", "4.6")]
```
Comment (franramirez688) [inline reference/conanfile/methods.rst:1271]: Perhaps, it's worth mentioning:

* The format of the return. List of different dict of settings? Does each dict have only one field?
* Are `options` allowed?

**Review by franramirez688 — APPROVED** [review]
I think it looks good! Anyway, have a look at my comment below to discuss if it could be necessary to add more details about the function.

**Comment by SSE4** [comment]
I think it deserves some further explanation.
as I can see, there are now, at least 3 different ways to express package compatibility:
1. [package_id/defining ABI compatibility](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html)
```

    def package_id(self):
        if self.settings.compiler.cppstd in ["14", "17"]
            self.info.settings.compiler.cppstd = "14/17"
```
2. [compatible packages](https://docs.conan.io/en/latest/reference/conanfile/methods.html#compatible-packages)
```
    def package_id(self):
        if self.settings.compiler == "gcc" and self.settings.compiler.cppstd == "14":
            compatible_pkg = self.info.clone()
            compatible_pkg.settings.compiler.cppstd = "17"
            self.compatible_packages.append(compatible_pkg)
```
3. [compatibility method](https://github.com/conan-io/conan/pull/10868)
```
    def compatibility(self):
         if self.settings.compiler == "gcc" and self.settings.compiler.cppstd == "14":
            return [{"settings": [("compiler.cppstd", "17")]]
```

for me personally, and I believe for many readers, the following points are still unclear:
- is any of them recommended over others? if I have a choice, which approach should I prefer and why?
- are they mutually exclusive, or could be used simultaneously in the same recipe?
- if all 3 of them are defined in the same recipe, but in different ways, which one takes a priority during the actual graph resolution?

**Review by franramirez688 — APPROVED** [review]

**Comment by czoido** [comment]
@SSE4,

I have extended the explanations and I think now it should be clear enough. Regarding your comments

> I think it deserves some further explanation. as I can see, there are now, at least 3 different ways to express package compatibility:
> 
> 1. [package_id/defining ABI compatibility](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html)
> 2. [compatible packages](https://docs.conan.io/en/latest/reference/conanfile/methods.html#compatible-packages)
> 3. [compatibility method](https://github.com/conan-io/conan/pull/10868)
> * is any of them recommended over others? if I have a choice, which approach should I prefer and why?
> * are they mutually exclusive, or could be used simultaneously in the same recipe?
> * if all 3 of them are defined in the same recipe, but in different ways, which one takes a priority during the actual graph resolution?

The difference between 1 and 2 is already explained in the docs in the introduction for [compatible packages](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#compatible-packages). They are different mechanisms, so now, we migrate from 2 to 3 that are equivalents.

Then it would not make sense to have both mechanisms (2 and 3) in a recipe, the docs now clearly state that you should move from one to the other.

Please, let me know if you have more questions, I'm merging this as we have to release 1.52 but we could improve this in another PR in the future. Thanks a lot for the feedback!

**Pull request merged**